### PR TITLE
[Avatar, Image] Add graceful fallbacks when fetching

### PR DESCRIPTION
**Pull request by kyledurand**
### WHY are these changes introduced?

Fixes #0000 

https://user-images.githubusercontent.com/6844391/162325154-04a5041e-7835-4c8e-a128-14dc7e261584.mp4

https://user-images.githubusercontent.com/6844391/162325137-d71b5562-dc04-4f90-af5b-70c5bcac633d.mp4

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';

import {
  Thumbnail,
  Card,
  MediaCard,
  Image,
  Layout,
  Page,
  Avatar,
  Stack,
} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      <Layout>
        <Layout.Section secondary>
          <MediaCard
            title="Media"
            description="description"
            primaryAction={{content: 'Primary action'}}
            secondaryAction={{content: 'Secondary action'}}
            popoverActions={[{content: 'Popover actions'}]}
            portrait
          >
            <Image
              alt=""
              source="https://images.unsplash.com/photo-1605007493699-af65834f8a00?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=3871&q=80"
              width="100%"
              height="100%"
              style={{objectFit: 'cover', objectPosition: 'center'}}
            />
          </MediaCard>
          <MediaCard
            title="Media"
            description="description"
            popoverActions={[{content: 'Popover actions'}]}
          >
            <Image
              alt=""
              source="https://images.unsplash.com/photo-1614102073832-030967418971?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=3869&q=80"
              width="100%"
              height="100%"
              style={{objectFit: 'cover', objectPosition: 'center'}}
            />
          </MediaCard>
          <MediaCard
            title="Media"
            description="description"
            popoverActions={[{content: 'Popover actions'}]}
          >
            <Image
              alt=""
              source="https://images.unsplash.com/photo-1614102073832-030967418971?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=3869&q=80"
              width="100%"
              height="100%"
              noPlaceholder
              style={{objectFit: 'cover', objectPosition: 'center'}}
            />
          </MediaCard>
        </Layout.Section>
        <Layout.Section>
          <Card sectioned>
            <Stack vertical>
              <Avatar
                source="https://images.unsplash.com/photo-1649186020589-21375439159a?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=4000&q=80"
                initials="KD"
                customer
              />
              <Avatar
                source="https://images.unsplash.com/photo-1592685530128-2f32bf8fe0fb?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=4480&q=80"
                initials="KD"
                customer
                name="Karl"
              />
              <Avatar source="image/source" />
              <Avatar source="image/source" initials="KD" name="KARL" />
              <Avatar customer name="KARL" />
              <Avatar initials="JO" />
              <Avatar initials="DW" />
              <Thumbnail
                alt=""
                source="https://images.unsplash.com/photo-1605007493699-af65834f8a00?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=3871&q=80"
              />
              <Thumbnail
                alt=""
                source="https://images.unsplash.com/photo-1614102073832-030967418971?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=3869&q=80"
              />
            </Stack>
          </Card>
        </Layout.Section>
      </Layout>
    </Page>
  );
}

```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
- [ ] For visual design changes, ping @ sarahill to update the Polaris UI kit

### CODE REVIEW

File: polaris-react/src/components/Avatar/Avatar.tsx [hunk 1]
```diff
@@ -95,24 +95,25 @@ export function Avatar({
...
         source={source}
         alt=""
         role="presentation"
         onLoad={handleLoad}
         onError={handleError}
       />
     ) : null;
-
   // Use `dominant-baseline: central` instead of `dy` when Edge supports it.
   const verticalOffset = '0.35em';
```
Comment (kyledurand) [inline polaris-react/src/components/Avatar/Avatar.tsx:118]: I played around with removing this in browserstack but it's still an issue in edge 101 😞

File: polaris-react/src/components/Image/Image.tsx [hunk 2]
```diff
@@ -1,38 +1,66 @@
...
 
 type CrossOrigin = 'anonymous' | 'use-credentials' | '' | undefined;
+type Status = 'loading' | 'loaded';
 
 export interface ImageProps extends React.HTMLProps<HTMLImageElement> {
   alt: string;
   source: string;
   crossOrigin?: CrossOrigin;
   sourceSet?: SourceSet[];
+  noPlaceholder?: boolean;
```
Comment (kyledurand) [inline polaris-react/src/components/Image/Image.tsx]: Maybe seeing a ton of gray blocks on an image heavy page isn't the right thing to do. Offering an escape hatch

File: polaris-react/src/components/Image/Image.tsx [hunk 3]
```diff
@@ -1,38 +1,66 @@
...
+  const [status, setStatus] = React.useState<Status>('loading');
   const finalSourceSet = sourceSet
     ? sourceSet
         .map(({source: subSource, descriptor}) => `${subSource} ${descriptor}`)
         .join(',')
     : null;
 
-  return finalSourceSet ? (
-    // eslint-disable-next-line jsx-a11y/alt-text
+  useEffect(() => setStatus('loading'), [source, sourceSet]);
```
Comment (kyledurand) [inline polaris-react/src/components/Image/Image.tsx:40]: Update loading status if source / sourceset changes

File: polaris-react/src/components/Image/Image.tsx [hunk 4]
```diff
@@ -1,38 +1,66 @@
...
+
+  return (
     <img
+      alt={alt}
       src={source}
-      srcSet={finalSourceSet}
       crossOrigin={crossOrigin}
+      className={className}
+      onLoad={handleLoad}
+      {...(finalSourceSet ? {srcSet: finalSourceSet} : {})}
```
Comment (kyledurand) [inline polaris-react/src/components/Image/Image.tsx:60]: This component was confusing. It was returning two `<img` elements based on props. This simplifies that by conditionally rendering the props instead

File: polaris-react/src/components/Image/Image.tsx [hunk 5]
```diff
@@ -1,38 +1,66 @@
...
       {...rest}
     />
-  ) : (
-    // eslint-disable-next-line jsx-a11y/alt-text
```
Comment (kyledurand) [inline polaris-react/src/components/Image/Image.tsx:35]: Also able to get rid of this eslint warning by just returning the one `img`

File: polaris-react/src/components/TopBar/README.md [hunk 6]
```diff
@@ -267,18 +267,12 @@ function TopBarExample() {
   const [isUserMenuOpen, setIsUserMenuOpen] = useState(false);
   const [isSearchActive, setIsSearchActive] = useState(false);
   const [searchValue, setSearchValue] = useState('');
-  const [isDarkTheme, setIsDarkTheme] = useState(false);
```
Comment (kyledurand) [inline polaris-react/src/components/TopBar/README.md:270]: Simplifying this example by removing the dark / light theme toggle which doesn't really work right now

File: polaris-react/src/components/Image/Image.tsx (hunk 2 above)
Comment (alex-page) [reply polaris-react/src/components/Image/Image.tsx]: Can we not add the escape hatch and see if this is a problem?

Comment (kyledurand) [reply polaris-react/src/components/Image/Image.tsx]: Knew you'd say that 😄

Comment (alex-page) [reply polaris-react/src/components/Image/Image.tsx]: Well it's hard to remove the escape hatch. I would love to get feedback or see in spin this being a problem.

Comment (kyledurand) [reply polaris-react/src/components/Image/Image.tsx]: Yep sounds good. I'll remove

**Review by alex-page — APPROVED** [review]

File: polaris-react/src/components/Image/Image.tsx [hunk 7]
```diff
@@ -16,23 +21,44 @@ export interface ImageProps extends React.HTMLProps<HTMLImageElement> {
   onError?(): void;
 }
 
-export function Image({sourceSet, source, crossOrigin, ...rest}: ImageProps) {
+export function Image({
+  alt,
```
Comment (alex-page) [inline polaris-react/src/components/Image/Image.tsx:25]: How was this missing 🤦

Comment (kyledurand) [reply polaris-react/src/components/Image/Image.tsx:25]: It wasn't actually I'm just exposing it. It was a required prop, just bundled up in `...rest`. Exposing it means we don't have to disable the a11y lint rule

Comment (alex-page) [reply polaris-react/src/components/Image/Image.tsx:25]: I see. I like getting it out of the rest 💯

**Pull request merged**